src/FSharpPlus/DList.fs

```diff
@@ -1,45 +1,260 @@
+        DList<'T>.foldBack f l state
+
+    let fold (f : ('State -> 'T -> 'State)) (state : 'State) (l:DList<'T>) =
+        DList<'T>.fold f state l
+
+    ///O(log n). Returns the first element.
+    let inline head (l:DList<'T>) = l.Head
+
+    ///O(log n). Returns option first element.
+    let inline tryHead (l:DList<'T>) = l.TryHead
+
+    ///O(1). Returns true if the DList has no elements.
+    let inline isEmpty (l:DList<'T>) = l.IsEmpty
+
+    ///O(1). Returns the count of elememts.
+    let inline length (l:DList<'T>) = l.Length
+
+    ///O(1). Returns DList of one elements.
+    let singleton x = DList(1, Unit x )
+
+    ///O(1). Returns a new DList with the element added to the end.
+    let inline conj x (l:DList<'T>) = l.Conj x
+
+    ///O(log n). Returns a new DList of the elements trailing the first element.
+    let inline tail (l:DList<'T>) = l.Tail
+
+    ///O(log n). Returns option DList of the elements trailing the first element.
+    let inline tryTail (l:DList<'T>) = l.TryTail
+
+    ///O(log n). Returns the first element and tail.
+    let inline uncons (l:DList<'T>) = l.Uncons
+
+    ///O(log n). Returns option first element and tail.
+    let inline tryUncons (l:DList<'T>) = l.TryUncons
+
+    ///O(n). Returns a DList of the seq.
+    let ofSeq s = DList<'T>.ofSeq s
+
+    ///O(n). Returns a list of the DList elements.
+    let inline toList l = foldBack (List.cons) l [] 
+
+    ///O(n). Returns a seq of the DList elements.
+    let inline toSeq (l:DList<'T>) = l :> seq<'T>
+
+    // additions to fit f#+ :
+    let inline map f (x:DList<_>)    = DList.foldBack (cons << f ) x empty
+    let concat x = DList.fold append empty x 
+    let inline join (f:DList<DList<_>>) = concat f
+    let inline ap f x = join <| map (fun y -> map ((|>) y) f) x
+    let inline bind m k              = DList.foldBack (append << k) empty m
 
 type DList<'T> with
     
-    static member get_Zero = DList id
-    static member (+) (DList x, DList y) = DList (x << y)
+    static member get_Zero = DList( 0, Nil)
+    static member (+) (x:DList<_>, y:DList<_>) = DList.append x y
 
-    static member get_Empty = DList id
-    static member (<|>) (DList x, DList y) = DList (x << y)
+    static member get_Empty = DList( 0, Nil)
+    static member (<|>) (x:DList<_>, y:DList<_>) = DList.append x y
     
-    static member ToSeq  x = toSeq  x
-    static member ToList x = toList x
-    static member OfSeq  x = ofSeq  x
-    static member OfList x = ofList x
-    static member Fold (x, f, z, [<Optional>]_impl) = fold f x z    
-
-    static member Return x = DList (fun r -> x::r)
-    static member Map (x, f) = map f x
-    static member (<*>) (f, x) = ap f x
-    static member Join x = join x
-    static member (>>=) (x, f) = bind x f
\ No newline at end of file
+    static member ToSeq  x = DList.toSeq  x
+    static member ToList x = DList.toList x
+    static member OfSeq  x = DList.ofSeq  x
```

Comment:
Shouldn't there be a OfList ?

Reply:
Well, in theory it's not really required for foldables, ``ofList`` function would work if it's a foldable, which requires ``OfSeq``.

But the situation today is that ``OfList`` doesn't have defaults defined, so it won't work.

I would say leave it off for now, once I create the defaults for ``OfList`` we can add it.

Reply:
Sounds good.